cmd/helm/plugins.go

```diff
@@ -0,0 +1,184 @@
+	if plugdirs == "" {
+		plugdirs = home.Plugins()
+	}
+	found := findPlugins(plugdirs)
+
+	// Now we create commands for all of these.
+	for _, path := range found {
+
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			// What should we do here? For now, write a message.
+			fmt.Fprintf(os.Stderr, "Failed to load %s: %s (skipping)", path, err)
+			continue
+		}
+
+		plug := &Plugin{}
+		if err := yaml.Unmarshal(data, plug); err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to parse %s: %s (skipping)", path, err)
+			continue
+		}
+
+		base := filepath.Base(path)
+		if plug.Usage == "" {
+			plug.Usage = fmt.Sprintf("the %q plugin", base)
+		}
+
+		c := &cobra.Command{
+			Use:   plug.Name,
+			Short: plug.Usage,
+			Long:  plug.Description,
+			RunE: func(cmd *cobra.Command, args []string) error {
+				setupEnv(plug.Name, base, plugdirs, home)
+				main, argv := plug.prepareCommand(args)
+				prog := exec.Command(main, argv...)
+				prog.Stdout = out
+				return prog.Run()
+			},
+			// This passes all the flags to the subcommand.
+			DisableFlagParsing: true,
+		}
+
+		if plug.UseTunnel {
+			c.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+				// Parse the parent flag, but not the local flags.
+				c.Parent().ParseFlags(args)
+				return setupConnection(cmd, args)
+			}
+		}
+
+		// TODO: Make sure a command with this name does not already exist.
+		baseCmd.AddCommand(c)
+	}
+}
+
+func (p *Plugin) prepareCommand(extraArgs []string) (string, []string) {
+	parts := strings.Split(os.ExpandEnv(p.Command), " ")
+	main := parts[0]
+	baseArgs := []string{}
+	if len(parts) > 1 {
+		baseArgs = parts[1:]
+	}
+	if !p.IgnoreFlags {
+		baseArgs = append(baseArgs, extraArgs...)
+	}
+	return main, baseArgs
+}
+
+// findPlugins returns a list of YAML files that describe plugins.
+func findPlugins(plugdirs string) []string {
+	found := []string{}
+	// Let's get all UNIXy and allow path separators
+	for _, p := range filepath.SplitList(plugdirs) {
+		// Look for any commands that match the Helm plugin pattern:
+		matches, err := filepath.Glob(filepath.Join(p, "*.yaml"))
+		if err != nil {
+			continue
+		}
+		found = append(found, matches...)
+	}
+	return found
+}
+
+func setupEnv(shortname, base, plugdirs string, home helmpath.Home) {
+	// Set extra env vars:
+	for key, val := range map[string]string{
+		"HELM_PLUGIN_SHORTNAME": shortname,
+		"HELM_PLUGIN_NAME":      base,
+		"HELM_BIN":              os.Args[0],
+
+		// Set vars that may not have been set, and save client the
+		// trouble of re-parsing.
+		pluginEnvVar: plugdirs,
+		homeEnvVar:   home.String(),
+
+		// Set vars that convey common information.
+		"HELM_PATH_REPOSITORY":       home.Repository(),
+		"HELM_PATH_REPOSITORY_FILE":  home.RepositoryFile(),
+		"HELM_PATH_CACHE":            home.Cache(),
+		"HELM_PATH_LOCAL_REPOSITORY": home.LocalRepository(),
+
+		// TODO: Add helm starter packs var when that is merged.
+	} {
+		os.Setenv(key, val)
```

Comment:
The reason is because Helm itself has to use those environment vars before it passes execution off to the plugin. I fixed things in the new code to make it more explicit.